crypto/merklekeystore/keystore.go

```diff
@@ -0,0 +1,295 @@
+	}
+
+	return Signature{
+		ByteSignature: signingKey.Sign(hashable),
+		Proof:         Proof(*proof),
+		VerifyingKey:  *signingKey.GetVerifyingKey(),
+	}, nil
+}
+
+// expects valid rounds, i.e round that are bigger than FirstValid.
+func (s *Signer) getMerkleTreeIndex(round uint64) uint64 {
+	return roundToIndex(s.FirstValid, round, s.Interval)
+}
+
+func (s *Signer) getArrayIndex(round uint64) (uint64, error) {
+	if round < s.FirstValid {
+		return 0, errReceivedRoundIsBeforeFirst
+	}
+	pos := roundToIndex(s.FirstValid, round, s.Interval)
+	pos = pos - s.ArrayBase
+	if round%s.Interval != 0 {
+		return pos, errNonExistantKey
+	}
+	if pos >= uint64(len(s.SignatureAlgorithms)) {
+		return 0, errOutOfBounds
+	}
+	return pos, nil
+}
+
+// Trim shortness deletes keys that existed before a specific round,
+// will return an error for non existing keys/ out of bounds keys.
+// the output is a copy of the signer - which can be persisted.
+func (s *Signer) Trim(before uint64) (*Signer, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	pos, err := s.getArrayIndex(before)
+	if err != nil {
+		return nil, err
+	}
+	s.dropKeys(int(pos))
+
+	// the array base represents the leaf index that we can treat as first leaf after trimming the ephemeral keys array.
+	s.ArrayBase = s.getMerkleTreeIndex(before) + 1
+	cpy := s.copy()
+
+	// Swapping the keys (both of them are the same, but the one in cpy doesn't contain a dangling array behind it.
+	// e.g: A=A[len(A)-20:] doesn't mean the garbage collector will free parts of memory from the array.
+	// assuming that cpy will be used briefly and then dropped - it's better to swap their key slices.
+	s.SignatureAlgorithms, cpy.SignatureAlgorithms =
+		cpy.SignatureAlgorithms, s.SignatureAlgorithms
+
+	return cpy, nil
+}
+
+func (s *Signer) copy() *Signer {
+	signerCopy := Signer{
+		_struct: struct{}{},
+
+		SignatureAlgorithms: make([]crypto.SignatureAlgorithm, len(s.SignatureAlgorithms)),
+		FirstValid:          s.FirstValid,
+		Interval:            s.Interval,
+		ArrayBase:           s.ArrayBase,
+
+		Tree: s.Tree,
+		mu:   deadlock.RWMutex{},
+	}
+
+	copy(signerCopy.SignatureAlgorithms, s.SignatureAlgorithms)
+	return &signerCopy
+}
+
+func (s *Signer) dropKeys(upTo int) {
+	for i := 0; i <= upTo; i++ {
+		// zero the keys.
+		s.SignatureAlgorithms[i] = crypto.SignatureAlgorithm{}
+	}
+	s.SignatureAlgorithms = s.SignatureAlgorithms[upTo+1:]
+}
+
+// Verify receives a signature over a specific crypto.Hashable object, and makes certain the signature is correct.
+func (v *Verifier) Verify(firstValid, round, interval uint64, obj crypto.Hashable, sig Signature) error {
+	if !v.HasValidRoot {
+		return errCannotVerify
+	}
+	if firstValid == 0 {
+		firstValid = 1
+	}
+	if round < firstValid {
+		return errReceivedRoundIsBeforeFirst
+	}
+
+	ephkey := CommittablePublicKey{
+		VerifyingKey: sig.VerifyingKey,
+		Round:        round,
+	}
+
+	pos := roundToIndex(firstValid, round, interval)
+	isInTree := merklearray.Verify(
```

Comment:
please change this variable name to something like err since the return type for Verify is error.